toolkit/tools/internal/grub/grubtokenizer.go

```diff
@@ -0,0 +1,509 @@
+				t.scanner.Col())
+		}
+
+		switch c {
+		// End of string
+		case '"':
+			// Consume " char.
+			t.scanner.Next()
+			break loop
+
+		// Escape character
+		case '\\':
+			// Consume \ character.
+			t.scanner.Next()
+
+			c, eof := t.scanner.Peek()
+			if eof {
+				return fmt.Errorf("unexpected end-of-file after '\\' character (%d:%d)", t.scanner.Line(),
+					t.scanner.Col())
+			}
+
+			switch c {
+			// Within double-quoted strings, only some characters are valid escape sequences.
+			case '$', '"', '\\':
+				sb.WriteRune(c)
+
+			case '\n':
+				// Drop the escaped newline.
+
+			default:
+				// Invalid escape sequences preserve the '\' character.
+				sb.WriteRune('\\')
+				sb.WriteRune(c)
+			}
+
+			t.scanner.Next()
+
+		// Variable expansion
+		case '$':
+			// Close out the current string token.
+			locEnd := t.scanner.Loc()
+			if locEnd.Index != locStart.Index {
+				subWord := t.newSubWord(locStart, locEnd, STRING, sb.String())
+				t.subWords = append(t.subWords, subWord)
+			}
+
+			// Parse the variable expansion.
+			err := t.parseVariableExpansion(QUOTED_VAR_EXPANSION)
+			if err != nil {
+				return err
+			}
+
+			// Restart parsing the double-quoted string.
+			locStart = t.scanner.Loc()
+			sb = strings.Builder{}
+
+		// Normal character
+		default:
+			sb.WriteRune(c)
+			t.scanner.Next()
+		}
+	}
+
+	subWord := t.newSubWord(locStart, t.scanner.Loc(), STRING, sb.String())
+	t.subWords = append(t.subWords, subWord)
+	return nil
+}
+
+// Parses single quoted string.
+// Single quoted strings don't accept any character escapes and end at the first ' character.
+func (t *grubConfigTokenizer) parseSingleQuotedString() error {
+	locStart := t.scanner.Loc()
+
+	// Consume ' char.
+	t.scanner.Next()
+
+	sb := strings.Builder{}
+loop:
+	for {
+		c, eof := t.scanner.Peek()
+		if eof {
+			return fmt.Errorf("unexpected end-of-file during single-quoted string (%d:%d)", t.scanner.Line(),
+				t.scanner.Col())
+		}
+
+		switch c {
+		// End of string
+		case '\'':
+			// Consume ' char.
+			t.scanner.Next()
+			break loop
+
+		// Normal character
+		default:
+			sb.WriteRune(c)
+			t.scanner.Next()
+		}
+	}
+
+	subWord := t.newSubWord(locStart, t.scanner.Loc(), STRING, sb.String())
+	t.subWords = append(t.subWords, subWord)
+	return nil
+}
+
+// Parses variable expansions.
+// e.g. $a, ${a}
+func (t *grubConfigTokenizer) parseVariableExpansion(subWordType SubWordType) error {
+	locStart := t.scanner.Loc()
+
+	// Consume $ char.
+	t.scanner.Next()
+
+	// Check if name is surrounded by braces.
+	c, eof := t.scanner.Peek()
+	usesBraces := false
+	if !eof && c == '{' {
+		usesBraces = true
+		// Consume { char
+		t.scanner.Next()
+	}
+
+	sb := strings.Builder{}
+loop:
+	for {
+		c, eof := t.scanner.Peek()
+		if eof {
+			break loop
+		}
+
+		switch {
+		// Name characters.
+		case (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9') || c == '_':
```

Comment:
should that also include dots `.` ?

Reply:
Variable names in Bash/grub.cfg can't contain dots.